packages/logging_cloudwatch/aws_logging_cloudwatch/lib/src/cloudwatch_logger_plugin.dart

```diff
@@ -0,0 +1,316 @@
+        duration: plugin._pluginConfig.flushIntervalInSeconds,
+        callback: plugin._startSyncingIfNotInProgress,
+        onError: plugin._onTimerError,
+      );
+    }
+    return plugin;
+  }
+
+  /// An [AWSLoggerPlugin] to use only for testing.
+  @protected
+  @visibleForTesting
+  CloudWatchLoggerPlugin.testPlugin({
+    required CloudWatchLogsClient client,
+    required CloudWatchLoggerPluginConfiguration pluginConfig,
+    required CloudWatchLogStreamProvider logStreamProvider,
+    required QueuedItemStore logStore,
+    RemoteLoggingConstraintProvider? remoteLoggingConstraintProvider,
+  })  : _syncing = false,
+        _enabled = pluginConfig.enable,
+        _pluginConfig = pluginConfig,
+        _logStore = logStore,
+        _remoteLoggingConstraintProvider = remoteLoggingConstraintProvider,
+        _logStreamProvider = logStreamProvider,
+        _clinet = client;
+
+  CloudWatchLoggerPlugin._({
+    required AWSCredentialsProvider credentialsProvider,
+    required CloudWatchLoggerPluginConfiguration pluginConfig,
+    RemoteLoggingConstraintProvider? remoteLoggingConstraintProvider,
+    CloudWatchLogStreamProvider? logStreamProvider,
+    QueuedItemStore? logStore,
+  })  : _syncing = false,
+        _enabled = pluginConfig.enable,
+        _pluginConfig = pluginConfig,
+        _logStore = logStore ?? InMemoryQueuedItemStore(),
+        _remoteLoggingConstraintProvider = remoteLoggingConstraintProvider ??
+            (pluginConfig.defaultRemoteConfiguration != null
+                ? DefaultRemoteLoggingConstraintProvider(
+                    config: pluginConfig.defaultRemoteConfiguration!,
+                    credentialsProvider: credentialsProvider,
+                  )
+                : null),
+        _clinet = CloudWatchLogsClient(
+          region: pluginConfig.region,
+          credentialsProvider: credentialsProvider,
+        ),
+        _logStreamProvider = logStreamProvider ??
+            DefaultCloudWatchLogStreamProvider(
+              logGroupName: pluginConfig.logGroupName,
+              client: CloudWatchLogsClient(
+                region: pluginConfig.region,
+                credentialsProvider: credentialsProvider,
+              ),
+            );
+
+  final CloudWatchLoggerPluginConfiguration _pluginConfig;
+  final CloudWatchLogsClient _clinet;
+  final CloudWatchLogStreamProvider _logStreamProvider;
+  final QueuedItemStore _logStore;
+  bool _syncing;
+  bool _enabled;
+  StoppableTimer? _timer;
+  RemoteLoggingConstraintProvider? _remoteLoggingConstraintProvider;
+
+  set remoteLoggingConstraintProvider(
+    RemoteLoggingConstraintProvider remoteProvider,
+  ) {
+    if (_remoteLoggingConstraintProvider != null) {
+      logger.error(
+        'remoteLoggingConstraintProvider is already configured.',
+      );
+      return;
+    }
+    _remoteLoggingConstraintProvider = remoteProvider;
+  }
+
+  Future<void> _startSyncingIfNotInProgress() async {
```

Comment:
I've been reading on dart concurrency, event loop, mutex and etc. and so far my understanding is race condition should not happen bc dart runs it on a signle thread and the `_syncing` flag is used to run the async `_startSyncingIfNotInProgress()` only once on multiple calls. am i getting it right or missing something here?

Reply:
The check should be sufficient and I may have misread it before.

My concern is that this is not a robust design. Error states are not being handled adequately, there is no retry logic, and it's not clear to me the overall state machine. Because this class is, in fact, a state machine - it's just that the possible states are not clear and the logic to move between states is not easily traceable.